extension/parquet/column_writer.cpp

```diff
@@ -1150,24 +1201,169 @@ class StringColumnWriter : public BasicColumnWriter {
+		}
+	}
+
+	void WriteVector(Serializer &temp_writer, ColumnWriterStatistics *stats_p, ColumnWriterPageState *page_state_p,
 	                 Vector &input_column, idx_t chunk_start, idx_t chunk_end) override {
+		auto &page_state = (StringWriterPageState &)*page_state_p;
 		auto &mask = FlatVector::Validity(input_column);
 		auto &stats = (StringStatisticsState &)*stats_p;
 
 		auto *ptr = FlatVector::GetData<string_t>(input_column);
-		for (idx_t r = chunk_start; r < chunk_end; r++) {
-			if (mask.RowIsValid(r)) {
+		if (page_state.IsDictionaryEncoded()) {
+			// dictionary based page
+			for (idx_t r = chunk_start; r < chunk_end; r++) {
+				if (!mask.RowIsValid(r)) {
+					continue;
+				}
+				auto value_index = page_state.dictionary.at(ptr[r].GetString());
+				if (!page_state.written_value) {
+					// first value
+					// write the bit-width as a one-byte entry
+					temp_writer.Write<uint8_t>(page_state.bit_width);
+					// now begin writing the actual value
+					page_state.encoder.BeginWrite(temp_writer, value_index);
+					page_state.written_value = true;
+				} else {
+					page_state.encoder.WriteValue(temp_writer, value_index);
+				}
+			}
+		} else {
+			// plain page
+			for (idx_t r = chunk_start; r < chunk_end; r++) {
+				if (!mask.RowIsValid(r)) {
+					continue;
+				}
 				stats.Update(ptr[r]);
 				temp_writer.Write<uint32_t>(ptr[r].GetSize());
 				temp_writer.WriteData((const_data_ptr_t)ptr[r].GetDataUnsafe(), ptr[r].GetSize());
 			}
 		}
 	}
 
-	idx_t GetRowSize(Vector &vector, idx_t index) override {
-		auto strings = FlatVector::GetData<string_t>(vector);
-		return strings[index].GetSize();
+	unique_ptr<ColumnWriterPageState> InitializePageState(BasicColumnWriterState &state_p) override {
+		auto &state = (StringColumnWriterState &)state_p;
+		return make_unique<StringWriterPageState>(state.key_bit_width, state.dictionary);
+	}
+
+	void FlushPageState(Serializer &temp_writer, ColumnWriterPageState *state_p) override {
+		auto &page_state = (StringWriterPageState &)*state_p;
+		if (page_state.bit_width != 0) {
+			if (!page_state.written_value) {
+				// all values are null
+				// just write the bit width
+				temp_writer.Write<uint8_t>(page_state.bit_width);
+				return;
+			}
+			page_state.encoder.FinishWrite(temp_writer);
+		}
+	}
+
+	duckdb_parquet::format::Encoding::type GetEncoding(BasicColumnWriterState &state_p) override {
+		auto &state = (StringColumnWriterState &)state_p;
+		return state.IsDictionaryEncoded() ? Encoding::RLE_DICTIONARY : Encoding::PLAIN;
+	}
+
+	void FlushDictionary(BasicColumnWriterState &state_p, ColumnWriterStatistics *stats_p) override {
+		auto &stats = (StringStatisticsState &)*stats_p;
+		auto &state = (StringColumnWriterState &)state_p;
+		if (state.IsDictionaryEncoded()) {
+			// first we need to sort the values in index order
+			auto values = vector<string>(state.dictionary.size());
+			for (auto iter = state.dictionary.cbegin(); iter != state.dictionary.cend(); iter++) {
```

Comment:
Prefer modern C++-style loop: `for(auto &entry : state.dictionary)`

Reply:
sure. I originally wanted constant iterators (happy to hear how to do it modern c++ style ) - but now the loop is so small that I'm less concerned about it. Will switch.

Reply:
Here you go:

`for(const auto &entry : state.dictionary)`